onnxruntime/test/contrib_ops/bitmask_dropout_op_test.cc

```diff
@@ -0,0 +1,174 @@
+  }
+
+  if (has_training_mode) {
+    test.AddInput<bool>("training_mode", {}, {false});
+  }
+
+  test.AddOutput<T>("output", input_dims, input_data);
+  if (has_mask) {
+    size_t mask_size = (input_size + kNumBitsPerBitmaskElement - 1) / kNumBitsPerBitmaskElement;
+    std::vector<BitmaskElementType> mask_data(mask_size, 0xFFFFFFFF);
+    test.AddOutput<BitmaskElementType>("mask", {static_cast<int64_t>(mask_size)}, mask_data);
+  }
+
+  std::vector<std::unique_ptr<IExecutionProvider>> test_eps;
+#ifdef USE_CUDA
+  test_eps.emplace_back(DefaultCudaExecutionProvider());
+#elif USE_ROCM
+  test_eps.emplace_back(DefaultRocmExecutionProvider());
+#endif
+  test.Run(OpTester::ExpectResult::kExpectSuccess, "", {}, nullptr, &test_eps);
+}
+
+template <typename T>
+void RunTestForInferenceWrapper() {
+  RunTestForInference<T>({7, 9});
+  RunTestForInference<T>({7, 9}, true);
+  RunTestForInference<T>({2, 17}, false, true);
+  RunTestForInference<T>({2, 17}, true, true);
+  RunTestForInference<T>({2, 17}, true, true, true);
+
+  // Vectorized.
+  RunTestForInference<T>({4, 11});
+  RunTestForInference<T>({4, 11}, true);
+  RunTestForInference<T>({16, 16}, false, true);
+  RunTestForInference<T>({16, 16}, true, true);
+  RunTestForInference<T>({16, 16}, true, true, true);
+}
+
+std::vector<BitmaskElementType> MasksToBitmasks(size_t size, const bool* mask_data) {
+  std::vector<BitmaskElementType> result;
+  for (size_t i = 0; i < size; ++i) {
+    size_t bitmask_idx = i / kNumBitsPerBitmaskElement;
+    size_t bitmask_shift = i % kNumBitsPerBitmaskElement;
+    if (bitmask_idx >= result.size()) {
+      result.emplace_back(0);
+    }
+
+    if (mask_data[i]) {
+      result[bitmask_idx] |= (1 << bitmask_shift);
+    }
+  }
+
+  return result;
+}
+
+template <typename T>
+void RunTestForTraining(const std::vector<int64_t>& input_dims) {
+  size_t input_size =
+      static_cast<size_t>(std::accumulate(input_dims.begin(), input_dims.end(), 1LL, std::multiplies<int64_t>()));
+  std::vector<T> input_data = ValueRange<T>(input_size, T(0.f), T(1.f));
+  for (float ratio : kRatios) {
+    OpTester dropout("Dropout", 13, kOnnxDomain, false);
+    dropout.AddAttribute<int64_t>("seed", kMaskSeed);
+    dropout.AddInput<T>("data", input_dims, input_data);
+    dropout.AddInput<float>("ratio", {}, {ratio});
+    dropout.AddInput<bool>("training_mode", {}, {true});
+
+    // Outputs need to be populated with a value (even if incorrect), Test output won't actually be checked.
+    // OpTester has no way to populate an output while explicitly ignoring its value.
+    // We will fetch the outputs manually and use them to verify BitmaskDropout outputs.
+    std::unique_ptr<bool[]> mask_buffer = std::make_unique<bool[]>(input_size);
+    dropout.AddOutput<T>("output", input_dims, input_data);
+    dropout.AddOutput<bool>("mask", input_dims, mask_buffer.get(), input_size);
+
+    std::vector<std::unique_ptr<IExecutionProvider>> dropout_eps;
+#ifdef USE_CUDA
```

Comment:
So we should put the test in "#ifdef USE_CUDA || USE_ROCM" ?

Reply:
I mean is this test workable without those two EPs.

Reply:
I mean is this test workable without those two EPs.

Reply:
At least I know it would fail for TRT. Actually these kernels are implemented for CUDA/ROCm only (even no CPU implied), so just test them on CUDA/ROCm.